File: apple-client/Hearth/Core/Sources/HearthCore/Cards/CardStore.swift
//
//  CardStore.swift
//  Hearth
//
//  Holds the generative UI cards driven by the Valar gateway's `ui_component`
//  messages. The op vocabulary is Echo's (`upsert` / `clear` / `clear_all`; an
//  absent op means upsert), but the RETENTION model is the desktop's as of
//  2026-07-31: cards are TRANSCRIPT HISTORY, not a status board. Every upsert
//  appends a NEW instance — a re-emit of a type never replaces or moves the
//  card the user already scrolled past. The list is capped at `maxCards` to
//  bound memory. `clear` still drops every instance of a type.
//
//  Exception: `session_gallery` is a transient picker, not a transcript entry,
//  so it stays one-per-type (see `singletonTypes`).
//

import Foundation
import Combine

@MainActor
final class CardStore: ObservableObject {
    /// Card history, oldest first. Many instances per type are expected.
    @Published private(set) var cards: [UiComponentDescriptor] = []

    /// Matches the desktop's `slice(-40)`.
    private static let maxCards = 40

    /// Types that are UI surfaces rather than transcript entries: a second
    /// emit replaces the first instead of stacking a duplicate.
    private static let singletonTypes: Set<String> = [
        UiComponentDescriptor.typeSessionGallery
    ]

    /// Pending expiry tasks, keyed by card INSTANCE id.
    private var ttlTasks: [String: Task<Void, Never>] = [:]

    // MARK: - Apply a raw ui_component payload

    func apply(_ raw: [String: Any]) {
        let op = raw.optString("op", fallback: "upsert")

        switch op {
        case "clear":
            let type = (raw["type"] as? String)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !type.isEmpty else { return }
            dismissType(type)
            print("[UI] clear type=\(type)")
            return

        case "clear_all":
            cancelAllTtl()
            cards.removeAll()
            print("[UI] clear_all")
            return

        case "upsert":
            break

        default:
            // Unknown op: ignore (forward compatibility).
            return
        }

        guard let descriptor = UiComponentDescriptor.from(raw) else {
            print("[UI] ui_component payload unusable (bad type or version); ignored")
            return
        }

        if Self.singletonTypes.contains(descriptor.type) {
            dismissType(descriptor.type)
        }

        cards.append(descriptor)
        if cards.count > Self.maxCards {
            let dropped = cards.prefix(cards.count - Self.maxCards)
            for card in dropped { cancelTtl(for: card.id) }
            cards.removeFirst(cards.count - Self.maxCards)
        }
        print("[UI] append type=\(descriptor.type) v=\(descriptor.version) (\(cards.count) in feed)")

        scheduleTtlIfNeeded(raw, id: descriptor.id, type: descriptor.type)
    }

    func clearAll() {
        cancelAllTtl()
        cards.removeAll()
    }

    /// Manually dismiss a single card instance (close button, gallery pick).
    func dismiss(_ id: UiComponentDescriptor.ID) {
        cancelTtl(for: id)
        cards.removeAll { $0.id == id }
    }

    /// Drop every instance of a type (the `clear` op, and singleton replacement).
    func dismissType(_ type: String) {
        for card in cards where card.type == type { cancelTtl(for: card.id) }
        cards.removeAll { $0.type == type }
    }

    // MARK: - TTL expiry

    private func scheduleTtlIfNeeded(_ raw: [String: Any], id: String, type: String) {
        guard let ttl = ttlSeconds(raw), ttl > 0 else { return }
        let nanos = UInt64(ttl * 1_000_000_000)
        ttlTasks[id] = Task { [weak self] in
            try? await Task.sleep(nanoseconds: nanos)
            guard !Task.isCancelled else { return }
            await MainActor.run {
                guard let self else { return }
                self.cards.removeAll { $0.id == id }
                self.ttlTasks[id] = nil
                print("[UI] expired type=\(type) (ttl=\(ttl)s)")
            }
        }
    }

    /// `ttl_s` is a top-level field; tolerate number or numeric string.
    private func ttlSeconds(_ raw: [String: Any]) -> Double? {
        if let d = raw["ttl_s"] as? Double { return d }
        if let i = raw["ttl_s"] as? Int { return Double(i) }
        if let s = raw["ttl_s"] as? String { return Double(s) }
        return nil
    }

    private func cancelTtl(for id: String) {
        ttlTasks[id]?.cancel()
        ttlTasks[id] = nil
    }

    private func cancelAllTtl() {
        for task in ttlTasks.values { task.cancel() }
        ttlTasks.removeAll()
    }
}

File: android-client/core/src/main/java/com/hearth/core/cards/CardStore.kt
package com.hearth.core.cards

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.util.UUID

/**
 * One generative-UI card, as the house describes it. Ported from the iOS
 * `UiComponentDescriptor`.
 *
 * The id is unique PER INSTANCE, not per type. The feed is a transcript, so
 * two emits of the same type are two entries that both keep their place: the
 * desktop learned live that keying by type made the earlier card vanish and
 * the newer one appear to "move down".
 *
 * Decoding is tolerant by contract: unknown keys are ignored, missing keys
 * fall back, and only a missing type or an unspeakable version makes a
 * payload unusable. An unknown TYPE parses fine and renders nothing, which is
 * what lets the house ship a new card before the client learns it.
 */
data class CardDescriptor(
    val id: String,
    val type: String,
    val version: Int,
    val props: JSONObject,
    val receivedAt: Long,
) {
    // Defensive accessors. Props are strings end to end on the wire, but a
    // handler that sends a number must not blank the field.
    fun str(key: String, fallback: String = ""): String {
        if (!props.has(key) || props.isNull(key)) return fallback
        return props.optString(key, fallback).ifEmpty { fallback }
    }

    fun int(key: String, fallback: Int = 0): Int = props.optInt(key, fallback)

    fun long(key: String, fallback: Long = 0L): Long = props.optLong(key, fallback)

    fun objList(key: String): List<JSONObject> {
        val array = props.optJSONArray(key) ?: return emptyList()
        return (0 until array.length()).mapNotNull { array.optJSONObject(it) }
    }

    fun strList(key: String): List<String> {
        val array = props.optJSONArray(key) ?: return emptyList()
        return (0 until array.length()).mapNotNull { array.optString(it).ifEmpty { null } }
    }

    companion object {
        const val SUPPORTED_VERSION = 1

        // The server vocabulary, from backend tools/card_catalog.yaml.
        const val CLOCK = "clock"
        const val WEATHER = "weather_card"
        const val TIMER = "timer_card"
        const val BRIEF_TEXT = "brief_text"
        const val CAPTIONS = "captions"
        const val SLIDESHOW = "slideshow"
        const val GENERATED_VIEW = "generated_view"
        const val IMAGE = "image_card"
        const val TERMINAL = "terminal_card"
        const val PERMISSION = "permission_card"
        const val CHOICE = "choice_card"

        fun from(raw: JSONObject): CardDescriptor? {
            val type = raw.optString("type").trim()
            if (type.isEmpty()) return null
            val version = raw.optInt("version", SUPPORTED_VERSION)
            if (version != SUPPORTED_VERSION) return null
            return CardDescriptor(
                id = UUID.randomUUID().toString(),
                type = type,
                version = version,
                props = raw.optJSONObject("props") ?: JSONObject(),
                receivedAt = System.currentTimeMillis(),
            )
        }
    }
}

/**
 * The cards the house has drawn this session. Ported from the iOS
 * `CardStore`.
 *
 * The op vocabulary is the Echo client's (upsert, clear, clear_all; an absent
 * op means upsert), but the RETENTION model is the desktop's: cards are
 * TRANSCRIPT HISTORY, not a status board. Every upsert APPENDS a new
 * instance, and a re-emit of a type never replaces or moves a card the person
 * has already scrolled past. `clear` still drops every instance of a type.
 */
class CardStore(private val scope: CoroutineScope) {

    private val _cards = MutableStateFlow<List<CardDescriptor>>(emptyList())

    /** Card history, oldest first. Many instances per type are expected. */
    val cards: StateFlow<List<CardDescriptor>> = _cards.asStateFlow()

    /** Pending expiry jobs, keyed by card INSTANCE id. */
    private val ttlJobs = mutableMapOf<String, Job>()

    fun apply(raw: JSONObject) {
        when (val op = raw.optString("op").ifEmpty { "upsert" }) {
            "clear" -> {
                val type = raw.optString("type").trim()
                if (type.isNotEmpty()) dismissType(type)
                return
            }

            "clear_all" -> {
                clearAll()
                return
            }

            "upsert" -> Unit

            // Unknown op: ignore, for forward compatibility.
            else -> {
                Log.i(TAG, "unknown card op '$op'")
                return
            }
        }

        val card = CardDescriptor.from(raw) ?: run {
            Log.i(TAG, "card payload unusable (bad type or version); ignored")
            return
        }

        val next = _cards.value + card
        _cards.value =
            if (next.size > MAX_CARDS) {
                next.takeLast(MAX_CARDS).also { kept ->
                    val keptIds = kept.map { it.id }.toSet()
                    next.filterNot { it.id in keptIds }.forEach { cancelTtl(it.id) }
                }
            } else {
                next
            }

        Log.i(TAG, "append ${card.type} (${_cards.value.size} in feed)")
        scheduleTtl(raw, card)
    }

    fun clearAll() {
        ttlJobs.values.forEach { it.cancel() }
        ttlJobs.clear()
        _cards.value = emptyList()
    }

    /** Dismiss one instance: a close button, or a gallery pick. */
    fun dismiss(id: String) {
        cancelTtl(id)
        _cards.value = _cards.value.filterNot { it.id == id }
    }

    /** Drop every instance of a type: the `clear` op. */
    fun dismissType(type: String) {
        _cards.value.filter { it.type == type }.forEach { cancelTtl(it.id) }
        _cards.value = _cards.value.filterNot { it.type == type }
    }

    /** `ttl_s` is a top-level field; tolerate a number or a numeric string. */
    private fun scheduleTtl(raw: JSONObject, card: CardDescriptor) {
        val ttl = when {
            raw.has("ttl_s") -> raw.optDouble("ttl_s", 0.0)
            else -> 0.0
        }
        if (ttl <= 0) return
        ttlJobs[card.id] = scope.launch {
            delay((ttl * 1000).toLong())
            _cards.value = _cards.value.filterNot { it.id == card.id }
            ttlJobs.remove(card.id)
            Log.i(TAG, "expired ${card.type} (ttl=${ttl}s)")
        }
    }

    private fun cancelTtl(id: String) {
        ttlJobs.remove(id)?.cancel()
    }

    companion object {
        private const val TAG = "HearthCards"

        /** Matches the desktop's slice(-40). */
        private const val MAX_CARDS = 40
    }
}
